### PR TITLE
implement-cni-reconcilation

### PR DESCRIPTION
**Pull request by calvix**
implement reconciliation for aws CNI resources, the service will create:
VPCCidrAssociation for the cluster VPC
* subnets for the was cni
* apply proper ENI configs to the wc cluster API so aws-cni can use right subnets and security groups

I will also delete these resources during  cluster deletion

towards https://github.com/giantswarm/capa-aws-cni-operator/pull/8

### CODE REVIEW

**Comment by calvix** [comment]
opinions are welcome, this is the last of many iteration of it that actually works

File: controllers/awscluster_controller.go [hunk 1]
```diff
@@ -67,6 +75,30 @@ func (r *AWSClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
 
 	logger = logger.WithValues("cluster", clusterName)
 
+	if awsCluster.Spec.NetworkSpec.VPC.ID == "" {
+		logger.Info("AWSCluster do not have vpc id set yet")
```
Comment (anvddriesch) [inline controllers/awscluster_controller.go:79]:
```suggestion
		logger.Info("AWSCluster does not have vpc id set yet")
```

File: controllers/awscluster_controller.go [hunk 2]
```diff
@@ -67,6 +75,30 @@ func (r *AWSClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
...
+		return ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: time.Minute * 2,
+		}, nil
+	}
+
+	if len(awsCluster.Spec.NetworkSpec.Subnets.GetUniqueZones()) == 0 {
+		logger.Info("AWSCluster do not have subnets set yet")
```
Comment (anvddriesch) [inline controllers/awscluster_controller.go:87]:
```suggestion
		logger.Info("AWSCluster does not have subnets set yet")
```

File: controllers/awscluster_controller.go [hunk 3]
```diff
@@ -67,6 +75,30 @@ func (r *AWSClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
...
+		return ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: time.Minute * 2,
+		}, nil
+	}
+
+	if _, ok := awsCluster.Status.Network.SecurityGroups[key.CNINodeSecurityGroupName]; !ok {
+		logger.Info("AWSCluster do not have security group ready yet")
```
Comment (anvddriesch) [inline controllers/awscluster_controller.go:95]:
```suggestion
		logger.Info("AWSCluster does not have security group ready yet")
```

File: controllers/awscluster_controller.go [hunk 4]
```diff
@@ -87,49 +119,83 @@ func (r *AWSClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
...
+	cniSecurityGroupID := awsCluster.Status.Network.SecurityGroups[key.CNINodeSecurityGroupName].ID
+
 	var cniService *cni.CNIService
-	{
+	// config for the CNI service
+	config := cni.CNIConfig{
+		AWSSession:         awsClientSession,
+		ClusterName:        clusterName,
+		CNISecurityGroupID: cniSecurityGroupID,
+		CtrlClient:         nil,              // we only need wc k8s client for resource creation, we dont need it for deletion, when cluster is being deleted it might not be avaiable
```
Comment (anvddriesch) [inline controllers/awscluster_controller.go:130]: Why do we set it explicitly to `nil`? I think if it were me I'd just leave it out. (Not saying that is better, just curios)

**Review by anvddriesch — APPROVED** [review]
LGTM, just nitpicking a little bit.

File: controllers/awscluster_controller.go (hunk 4 above)
Comment (calvix) [reply controllers/awscluster_controller.go:130]: i mostly wanted to put the comment there to explain why we do not set it here but later

I can remove it

**Review by njuettner — APPROVED** [review]

File: controllers/awscluster_controller.go (hunk 4 above)
Comment (anvddriesch) [reply controllers/awscluster_controller.go:130]: Ah no that makes perfect sense! No need to remove it. :)

**Pull request merged**